Tessel get seeker option

Student007 opened the pull request:
`Tessel.get` and `Tessel.list` accept additional option
`opts.authorized` for filtering seekers result.

fixes https://github.com/tessel/t2-cli/issues/368

rwaldron commented on lib/controller.js:
```diff
@@ -113,7 +114,8 @@ Tessel.get = function(opts) {
       usb: opts.usb,
-      lan: opts.lan
+      lan: opts.lan,
+      authorized: opts.authorized || true
```
If `opts.authorized` is explicitly set to `false`, this expression will evaluate to `true`, ie. `false || 1; // 1`

Student007 replied:
@rwaldron :scream: ... you are right - any idea ?

Student007 replied:
maybe 
```
authorized: opts.authorized || "yes" // or "no"
```

johnnyman727 replied:
I recommend passing through the option as is. Then, inside the [seeker.start](https://github.com/tessel/t2-cli/blob/master/lib/discover.js#L26-L47) function, you can sanitize the option. I don't think you need should set it to true by default.

rwaldron replied:
```js
authorized: Boolean(opts.authorized),
```

Will make... 

```js

Boolean(true);
true

Boolean(1);
true

Boolean(false);
false

Boolean(undefined);
false

Boolean(null);
false

Boolean(0);
false
```

Student007 replied:
@rwaldron thanks for this idea - never used `Boolean()` before.
I am used to have three (3) states to work with: 
- undefined (by `x === "undefined"`)
- true (everything with a value what is neither "undefined" nor false `if (x)`)
- false (if it is false: `if (!x)` what only should match if `x === false` means it will **NOT** match if `x === 0`)

The stupid expression `opts.authorized || true` is really a bad case because the `||` is triggered by `"undefined"` and `false`

johnnyman727 commented on lib/controller.js:
```diff
@@ -26,7 +26,8 @@ Tessel.list = function(opts) {
       usb: opts.usb,
-      lan: opts.lan
+      lan: opts.lan,
+      authorized: undefined
```
I recommend setting it to `false` explicitly.

johnnyman727 commented on lib/discover.js:
```diff
@@ -50,6 +50,18 @@ TesselSeeker.prototype.start = function(opts) {
     var tessel = new Tessel(conn);
     var opening = tessel.connection.open()
       .then(function() {
+        if (tessel.lanConnection && opts.authorized !== undefined) {
```
I think what we should do here is just [change the Tessel initialization](https://github.com/tessel/t2-cli/blob/master/lib/tessel/tessel.js#L5) to have `authorized = true` by default if it is a USB Connection. That way this can just be a simple if statement:

```
if (opts.authorized && !tessel.connection.authorized) {
  return;
}
```

Student007 replied:
@johnnyman727 will do it tomorrow :sleeping:

rwaldron replied:
> I think what we should do here is just change the Tessel initialization to have authorized = true

+1—this is semantically sound.

Student007 replied:
:+1:  - did this.

Student007 commented:
@johnnyman727 @rwaldron 
Changed default value and also added a reasonable `opts.authorized` to
each command what calls `Tessel.get` or
`controller.standardTesselCommand`` - also the CASE of Tessel.js
`connection.connectionType === USB` defaults to  `connection.authorized
=== true`

rwaldron commented:
I just did smoke testing with an unauthorized T2 and this nicely prevents the `Unhandled rejection Error: Not connected` issue. 

Once [this is merged](https://github.com/Student007/t2-cli/pull/4), then I'm +1 on this PR.

Student007 commented:
@rwaldron @johnnyman727 my local tests pass - don't know why Travis-CI is angry ... :sleeping: now

johnnyman727 commented on lib/controller.js:
```diff
@@ -113,17 +117,25 @@ Tessel.get = function(opts) {
     var seeker = new discover.TesselSeeker().start(seekerOpts);
+    var noTessels = seekerOpts.authorized ?
+      'No Authorized Tessels Found.' :
+      'No Tessels Found.';
```
@Student007 can we try to make this cleaner? It's not good that this is copied and pasted in `Tessel.get` and `Tessel.list`.

rwaldron replied:
That was my patch ;)

I figured we could do a follow up, but might as well just store them in vars now

Student007 replied:
@rwaldron @johnnyman727 is in `Seeker.start` now - usable like this:
```
var seeker = new discover.TesselSeeker().start(seekerOpts);
    var noTessels = opts.authorized ?
      seeker.msg.noAuth :
      seeker.msg.auth;
```
Tests are fixed by adding `msg` to stub of `Seeker.start`

johnnyman727 commented on lib/controller.js:
```diff
@@ -497,6 +510,7 @@ controller.provisionTessel = function(opts) {
 };
 
 controller.deployScript = function(opts) {
+  opts.authorized = true;
```
It's not great that we have this `opts.authorized = true` all over the place. I understand that you didn't want to make it an option for each command b/c the user doesn't need access to it. But perhaps we can place this line in [`callControllWith`](https://github.com/tessel/t2-cli/blob/8848a73c1c0b0d27c2aa291e8390c2e0e4200b36/bin/tessel-2.js#L34)?

rwaldron replied:
I did a test earlier where I put it in put it in standardTesselCommand, but there's one place where it's a `false`, and I had other stuff to do so I didn't press it

Student007 replied:
this was the reason I also didn't put it in `standardTesselCommand`

Student007 replied:
the static way I did this at the moment is done with the purpose it will be changed by other PRs after deeper thinking about each command. For example - @johnnyman727 is right - some could be an option within the CLI `bin/tessel-2.js`

johnnyman727 commented on lib/tessel/tessel.js:
```diff
@@ -32,6 +32,7 @@ function Tessel(connection) {
     // Set the connection var so we have an abstract interface to relay comms
     switch (connection.connectionType) {
       case 'USB':
+        connection.authorized = true;
```
Nice!

Student007 commented:
:tada: Travis-CI error fixed also - before :tea: + :egg: (breakfast) :yum:

rwaldron commented:
@Student007 great work!

rwaldron commented:
Landed

Student007 commented:
:notes: :dancers: :notes: 
great teamwork!
:bow:
